### DISCUSSION TITLE
Dear all interested twinBASIC users...

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#836] [General]
I'm well aware from recent posts that there are several of the VB6er's itching to get moving their VB6 projects over to twinBASIC, but are just needing feature XYZ before being able to do so.  I get it.  But I would politely like to ask for a little more patience please.  Good things come to those that wait :)

We are making massive strides every day, but please appreciate that VB6er's are not the only ones looking at twinBASIC.  As such, I aim to provide a constant mix of bug-fixes, back-compat improvements, and new stuff.  Some weeks will be more focused towards back-compat (like last week: ActiveX Controls), and other weeks towards the new stuff (like this week: WebView2).  That's just how it goes on this project, and as frustrating as it might be for some of you, it's in the best interest of this project in order to try to keep people from *all* sides interested, which will in turn help the wheels keep turning.  It also helps from my perpsonal perspective, as it keeps things fresh and interesting from week to week.

From my point of view, it's not particularly helpful to be listing the missing features that you *personally need* across different issues and discussions, which is what is happening at the moment, since everybody has slightly different wants and needs.    I'm aware of all the missing features already, as there are already open issues for them in the majority of cases (and if there aren't, feel free to open individual, specific issues for tracking purposes).

Thank you as always for your support and patience.

**Comment by bclothier** [comment]
For the record, I think it's WAY impressive that you were able to put together not only a new IDE but also basic ActiveX control support in a matter of 2 or 3 months. I know that were it up to me, everybody'd be waiting 84 years. 😆 It is the slow and steady that wins the race.

**Reply by jpbro** [reply]
Only 84 years? That's not bad! I'd be in a race with the heat death of the universe.

**Reply by WaynePhillipsEA** [reply]
Thanks @bclothier, much appreciated.  Looking back, it was probably a risky move to make our own IDE, but I think it was a risk that is paying off.  

Slow and steady is good... but unfortunately my brain doesn't like it!!

**Comment by greatquux** [comment]
Thanks for this update. I had sort of gotten the idea from the roadmap, but it was important to make it clear.  Eventually, it should be (at least almost if not quite 100%) compatible with our VB6 projects, but it's also something you want to add new features to in order to continue advancing the language (we're missing out on almost 20 years after all!) and it's fun to do so.  But you recognize the way to get people to buy and use it is to make it compatible and then they can modernize from there.  We've been waiting decades, we can wait a little more, and this really is impressive.  I may give another month to the RAD Basic project but I think so far it's pretty clear I'll be buying a subscription to twinBASIC.

**Reply by WaynePhillipsEA** [reply]
Indeed, the roadmap doesn't cover everything, but hopefully it gives a rough overview of the areas where I'm concentrating the work at each quarter.

I can't say much about RAD Basic, but it would sure be nice to have you onboard :)

**Comment by jpbro** [comment]
The pace of advancement has been beyond impressive, and I have 100% faith that Wayne can get us to the promised land provided the business case can continue to be made. Everyone out there (with the means) who is looking for an eventual VB6 successor should absolutely subscribe to tB or at least [buy Wayne a coffee once a month](https://ko-fi.com/twinbasic). Keep up the fantastic work!

**Reply by WaynePhillipsEA** [reply]
Thanks for your support @jpbro!

**Comment by dmrvb** [comment]
For me the frustration is that it feels so very close, a while ago I posted a screenshot of my audio capture software running in real time in tB. All the stuff that I thought might be difficult, like the audio API calls, worked first time. I just need more GUI.   But I agree that tB is going to need to satisfy a range of users to suceed, and for me some new features/modernisation is ultimately more important than 100% compatability.

**Reply by WaynePhillipsEA** [reply]
Yup, I hear ya.  GUI is most definitely high priority, so watch this space.  A *lot* of effort went into matching VBx API handling (and all the quirks associated with it), so it's certainly nice to hear that it worked well for you.  We'll get there!

**Comment by Heartlander** [comment]
I'm an old fart, a coder from the Pleistocene Era. I've been coding in Access since it was introduced 30+ years ago. I am astonished at what you are doing here, Wayne. Did you wake up one morning, look out the window and say, "Hell, I think I'm gonna turn the RAD coding world on it's ear?" Amazing, indeed.

Couple things:  1. is says the Invite to the Discord is Invalid, or I'd be over there. 2. What happens if I take the VBA code from an Access project and compile it in tVB? How does it handle the intrinsic functions that Access itself provides, or does it? 3. I read your article on Security Through Obscurity. Are there plans for a Code Obfuscator to run prior to final Compile? I've done some work on something like that if it might help. I'm a FIRM BELIEVER in obfuscation in these modern times.

**Reply by WaynePhillipsEA** [reply]
Hey there @Heartlander, thanks for your support!

> Did you wake up one morning, look out the window and say, "Hell, I think I'm gonna turn the RAD coding world on it's ear?"

I could bore you with the long story of how I first contemplated the idea back in 2007 whilst relaxing on a beach in the Maldives, filling a notebook of ideas and topping up my tan, before ultimately deciding it was an outrageous idea and relegating the notebook to the attic.  But I won't ;)

1) Discord link looks OK, perhaps it was a temporary glitch.  Could you retry?
2) We're not emulating anything outside of the VB6/VBA standard libraries at the moment.  You can of course use Access as a COM reference and use the Access object model directly.
3) Are you referring to [this](https://nolongerset.com/security-through-obscurity/) article?  If so, that's by Mike Wolfe (@mwolfe02), but I do happen to agree with everything he says in that article.   Once we've got an [Analysis Extension API](https://github.com/WaynePhillipsEA/twinbasic/issues/51) I would be surprised if someone didn't write a tB code obfuscator with it.  It's not something that's on the tB roadmap (#335) at the moment as there are many other priorities, but I could certainly see us having something built-in eventually too.

**Comment by Heartlander** [comment]
It still says Invite Invalid.

RE: code obfuscation, I did some experiments with a little Access app I
wrote (see attached) and it did fine up to a point. THen, you realize that
you need a good search and replace capability to get in all the nooks and
crannies, something like  Philipp Stiefe's Find And Replace util
(CodeKabinett.com).I tried to interest him in exposing his code so we could
just pass him the strChangeFrom value and he pass back the strChangedTo
value but, for some reason, he doesn't seem to think obfuscation is worth
the effort. I think Luke Chung also had some chats with Phillip along the
same lines.

Do you talk with Luke and Phillip about tB? From my humble station as an
unwashed, uneducated, self-taught VB6/VBA coder, you guys are some of the
most knowledgeable people in the VB/Access realms. I'm not familiar with
Rubber Duck but am a firm believer in Luke's Access Analyzer. And so, yeah,
if the Fates should allow for tB to have such built in glories as a code
analyzer, search and replace with full final-compile code obfuscation, I
would be able to die a happy man, knowing the world was in good hands.


On Tue, Jun 14, 2022 at 1:22 PM WaynePhillipsEA ***@***.***>
wrote:

> Hey there @Heartlander <https://github.com/Heartlander>, thanks for your
> support!
>
> Did you wake up one morning, look out the window and say, "Hell, I think
> I'm gonna turn the RAD coding world on it's ear?"
>
> I could bore you with the long story of how I first contemplated the idea
> back in 2007 whilst relaxing on a beach in the Maldives, filling a notebook
> of ideas and topping up my tan, before ultimately deciding it was an
> outrageous idea and relegating the notebook to the attic. But I won't ;)
>
>    1. Discord link looks OK, perhaps it was a temporary glitch. Could you
>    retry?
>    2. We're not emulating anything outside of the VB6/VBA standard
>    libraries at the moment. You can of course use Access as a COM reference
>    and use the Access object model directly.
>    3. Are you referring to this
>    <https://nolongerset.com/security-through-obscurity/> article? If so,
>    that's by Mike Wolfe ***@***.*** <https://github.com/mwolfe02>), but I
>    do happen to agree with everything he says in that article. Once we've got
>    an Analysis Extension API
>    <https://github.com/WaynePhillipsEA/twinbasic/issues/51> I would be
>    surprised if someone didn't write a tB code obfuscator with it. It's not
>    something that's on the tB roadmap (#335
>    <https://github.com/WaynePhillipsEA/twinbasic/issues/335>) at the
>    moment as there are many other priorities, but I could certainly see us
>    having something built-in eventually too.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WaynePhillipsEA/twinbasic/discussions/836#discussioncomment-2949717>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGFI45A5M6QYPKW5WSDSFVTVPDEUPANCNFSM5VZCKSLA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
My PACred  ID:
VVZZEZEVZVPVZVZZEKXPPVELXTVVZVPVZVZZEVPZEVZVZVPETHXLZLZTZEAVPVZZEZEVZVPVZVZZEHHAXLAAALPPPTXLZXEAAKZXEVETXTXTZKEXXHKVPVZZEZEVZVPVZVZZEVPZELXLKTAHKVVLKXXVLVPVZZEZEVZVPVZVZZEVPZEVZZZAHAHAPZZXPKAZZEHAHKAXPLATXLAKZPALZT

**Reply by WaynePhillipsEA** [reply]
Where are you getting the bad discord link from?  The link in #629 is working (checked just now).

I used to be in regular contact with Luke, but we haven't spoken in a while and so I don't know if he knows of twinBASIC yet.  Philipp most definitely is aware of tB, as Mike had demonstrated tB twice at the Access DevCon 🙂

**Comment by Heartlander** [comment]
Wayne, I know you are up to your eyebrows in tB but has anyone sketched out
what all you would have to do to convert an Access app to a compiled
TwinBasic program? Thanks.

On Tue, Jun 14, 2022 at 2:40 PM WaynePhillipsEA ***@***.***>
wrote:

> Where are you getting the bad discord link from? The link in #629
> <https://github.com/WaynePhillipsEA/twinbasic/discussions/629> is working
> (checked just now).
>
> I used to be in regular contact with Luke, but we haven't spoken in a
> while and so I don't know if he knows of twinBASIC yet. Philipp most
> definitely is aware of tB, as Mike had demonstrated tB twice at the Access
> DevCon 🙂
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WaynePhillipsEA/twinbasic/discussions/836#discussioncomment-2950279>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGFI45FYWWKTJENFAGOUA7TVPDN2DANCNFSM5VZCKSLA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
My PACred  ID:
VVZZEZEVZVPVZVZZEKXPPVELXTVVZVPVZVZZEVPZEVZVZVPETHXLZLZTZEAVPVZZEZEVZVPVZVZZEHHAXLAAALPPPTXLZXEAAKZXEVETXTXTZKEXXHKVPVZZEZEVZVPVZVZZEVPZELXLKTAHKVVLKXXVLVPVZZEZEVZVPVZVZZEVPZEVZZZAHAHAPZZXPKAZZEHAHKAXPLATXLAKZPALZT

**Comment by Heartlander** [comment]
If I compile a VBA code module with a line of code that mentions an Access form data value ( like Forms!frm_Whatever!fld_Wherever.Value ) will the tB compiler hurl?

**Reply by Heartlander** [reply]
In the context of making a DLL, by the way.

**Reply by bclothier** [reply]
I think you should use a different syntax because you are no longer in a VBA project and thus those are now foreign objects to your DLL's context. 

Something like this:

```
'Assume CurrentApp is the Access.Application object returned from the OnConnect procedure of the IDTExtensibility interface

CurrentApp.Forms("frm_Whatever").Controls("fld_Wherever").Value
```

You should not use implicit `Access.Application` reference because that will leak. It's best to always fully qualify and work from the `Access.Application` instance returned to you when the [`OnConnect` procedure](https://docs.microsoft.com/en-us/dotnet/api/extensibility.idtextensibility2.onconnection?view=visualstudiosdk-2022) runs. If you've done automation from Access to Excel or Word, you probably have seen the issues where using the implicit reference to the `Application` objects cause a ghost process. The same thing will happen here if you use implicit reference to Access' `Application`.

**Reply by WaynePhillipsEA** [reply]
Just to add to @bclothier's answer... if you're not writing an Add-In, you'd typically pass the instance of the Access.Application object over to your DLL via a method call (e.g. `Set myClass = New MyTbClass : myClass.Init(Access.Application)`).

**Reply by WaynePhillipsEA** [reply]
BTW, this should've been a new discussion really.

**Comment by Heartlander** [comment]
Thank you, I appreciate your time. That makes perfect sense. 

Kent

**Comment by KimVes** [comment]
> 

I must say I'm really impressive what I have seen so far, very impressive!!! I have many years been thinking about a code renewal project from VB6. I have been looking in converting code to C# or rewrite everything from scratch. But now I think tB can be one option after I downloaded the beta 80 version and played around with it. 

Take you time and do everything properly. I can wait one/two more years before I make my final decisions how to proceed. But tB feels like a good option :-) Keep up the good work, and I will be following the progress...

I found following problems when I tested, with some features I know I need:
- LoadPicture cannot load png-files. But good that a png-files can be loaded in IDE (this is not possible in VB6)
- Transparent background in png-files is not displayed correctly
- Picturebox missing, with click event

But I understand that GUI is under constructions, so I hope these features also will be implemented :-)

**Comment by Heartlander** [comment]
Wayne, I feel bad taking up your time with this. When I got to that link
https://github.com/WaynePhillipsEA/twinbasic/discussions/629 I get:
[image: image.png]
Do I really need an Invite? Can you issue a new invite?



On Tue, Jun 14, 2022 at 2:40 PM WaynePhillipsEA ***@***.***>
wrote:

> Where are you getting the bad discord link from? The link in #629
> <https://github.com/WaynePhillipsEA/twinbasic/discussions/629> is working
> (checked just now).
>
> I used to be in regular contact with Luke, but we haven't spoken in a
> while and so I don't know if he knows of twinBASIC yet. Philipp most
> definitely is aware of tB, as Mike had demonstrated tB twice at the Access
> DevCon 🙂
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WaynePhillipsEA/twinbasic/discussions/836#discussioncomment-2950279>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGFI45FYWWKTJENFAGOUA7TVPDN2DANCNFSM5VZCKSLA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
My PACred  ID:
VVZZEZEVZVPVZVZZEKXPPVELXTVVZVPVZVZZEVPZEVZVZVPETHXLZLZTZEAVPVZZEZEVZVPVZVZZEHHAXLAAALPPPTXLZXEAAKZXEVETXTXTZKEXXHKVPVZZEZEVZVPVZVZZEVPZELXLKTAHKVVLKXXVLVPVZZEZEVZVPVZVZZEVPZEVZZZAHAHAPZZXPKAZZEHAHKAXPLATXLAKZPALZT

**Reply by WaynePhillipsEA** [reply]
Not sure why you're having problems with the original link as it it still valid and working for others.   Here's a new link for you to try:  https://discord.gg/xNgsw3Bm  (valid for 7 days)

**Comment by saturday220514** [comment]
![2022-10-12_22-16-21](https://user-images.githubusercontent.com/105542220/195449899-ee879344-8054-4c0d-8d4f-6388f3bcd589.jpg)

**Reply by WaynePhillipsEA** [reply]
Hi @saturday220514, is that with using the new link above, or the original link at the top?

**Reply by saturday220514** [reply]
Hi Wayne
it's the one from two days ago, it may just be affecting me but I was wondering if Heartlander is getting the same screen?

**Reply by WaynePhillipsEA** [reply]
For anyone having trouble joining our discord server, apparently creating a discord account and logging in to it before clicking our discord link should fix it.

https://discord.com/register
